### ISSUE TITLE
Custom logloss function has different behavior compared to original logloss.

### ISSUE THREAD

**Issue by hcz28** [Microsoft/LightGBM#2136]
I am trying to define some new objective functions in LightGBM for binary classification problem, so I tried the logloss first to see whether it has the same behavior compared to the original objective `objective: binary`. However, it seems that the results are different.

```python
import lightgbm as lgb
import numpy as np
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split

X, y = make_classification(n_samples=1000, n_features=10)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)
dtrain = lgb.Dataset(X_train, y_train)

# use original logloss
params = {
    'objective': 'binary',
    'n_estimators': 10,
    'verbose':0
}
model_1 = lgb.train(params=params, train_set=dtrain)
scores_1 = model_1.predict(X_test)

# use custom obj of logloss
def logloss(preds, data):
    y_true = data.get_label()
    preds = 1. / (1. + np.exp(-preds))
    grad = -(y_true - preds)
    hess = preds * (1.0 - preds)
    return grad, hess

model_2 = lgb.train(params=params, train_set=dtrain, fobj=logloss)
scores_2 = model_2.predict(X_test)
```

And here are the scores:
```
print scores_1[:20]

[ 0.18201168  0.67024934  0.81996772  0.21083916  0.79884412  0.81996772
  0.2840331   0.81996772  0.81996772  0.18234898  0.31901805  0.22919787
  0.32515892  0.81996772  0.80086571  0.23573144  0.17850571  0.41036328
  0.81996772  0.71983127]

print scores_2[:20]

[-1.49868901  0.61493815  1.52130631 -1.26545304  1.38389951  1.52130631
 -1.04258177  1.52130631  1.52130631 -1.49641211 -0.74349987 -1.08400048
 -0.75506528  1.52130631  1.39654386 -1.1947502  -1.52130631 -0.20657762
  1.52130631  0.80335815]

scores_2_trans = 1. / (1. + np.exp(-scores_2))

print scores_2_trans[:20]

[ 0.18262113  0.64906644  0.82073076  0.22003661  0.79961655  0.82073076
  0.26065215  0.82073076  0.82073076  0.18296126  0.32223929  0.25274971
  0.3197186   0.82073076  0.80163488  0.23241044  0.17926924  0.44853847
  0.82073076  0.69069236]
```

So `scores_1` and `scores_2_trans` are similar but not the same. Why is that?

Thanks!

## Environment info

Operating System: Mac OS

C++/Python/R version: Python 2.7

LightGBM version: 2.2.2

**Comment by guolinke** [comment]
firstly, there is the floating point sum-up error.
secondly, the custom loss is calculated by python, and built-in loss is calculated by cpp, it is hard to ensure the exact same results.
BTW, you should set 'boost_from_average=false', which is default true in 'binary', but false in custom loss.
